c/misra/src/rules/DIR-4-9/FunctionOverFunctionLikeMacro.ql

```diff
@@ -0,0 +1,102 @@
+/**
+ * @id c/misra/function-over-function-like-macro
+ * @name DIR-4-9: A function should be used in preference to a function-like macro where they are interchangeable
+ * @description Using a function-like macro instead of a function can lead to unexpected program
+ *              behaviour.
+ * @kind problem
+ * @precision medium
+ * @problem.severity recommendation
+ * @tags external/misra/id/dir-4-9
+ *       external/misra/audit
+ *       maintainability
+ *       readability
+ *       external/misra/obligation/advisory
+ */
+
+import cpp
+import codingstandards.c.misra
+import codingstandards.cpp.FunctionLikeMacro
+import codingstandards.cpp.Naming
+
+predicate omission(Macro i) { Naming::Cpp14::hasStandardLibraryMacroName(i.getName()) }
+
+abstract class IrreplaceableFunctionLikeMacro extends FunctionLikeMacro { }
+
+private class AsmArgumentInvoked extends IrreplaceableFunctionLikeMacro {
+  AsmArgumentInvoked() {
+    any(AsmStmt s).getLocation().subsumes(this.getAnInvocation().getLocation())
+  }
+}
+
+private class OnlyConstantNumericInvoked extends IrreplaceableFunctionLikeMacro {
+  OnlyConstantNumericInvoked() {
+    forex(MacroInvocation mi | mi = this.getAnInvocation() |
+      mi.getUnexpandedArgument(_).regexpMatch("\\d+")
+    )
+  }
+}
+
+private class KnownIrreplaceableFunctionLikeMacro extends IrreplaceableFunctionLikeMacro {
+  KnownIrreplaceableFunctionLikeMacro() {
+    this.getName() in ["UNUSED", "__has_builtin", "MIN", "MAX"]
+  }
+}
+
+private class UsedToStaticInitialize extends IrreplaceableFunctionLikeMacro {
+  UsedToStaticInitialize() {
+    any(StaticStorageDurationVariable v).getInitializer().getExpr() =
+      this.getAnInvocation().getExpr()
+  }
+}
+
+private class FunctionLikeMacroWithOperatorArgument extends IrreplaceableFunctionLikeMacro {
+  FunctionLikeMacroWithOperatorArgument() {
+    exists(MacroInvocation mi | mi.getMacro() = this |
+      mi.getUnexpandedArgument(_) = any(Operation op).getOperator()
+    )
+  }
+}
+
+abstract class UnsafeMacro extends FunctionLikeMacro { }
```

Comment:
We should be clear about what constitutes an unsafe macro.

Reply:
not relevant any longer